pkg/controller/hyperconverged/hyperconverged_controller.go

```diff
@@ -121,325 +121,115 @@ func (r *ReconcileHyperConverged) Reconcile(request reconcile.Request) (reconcil
-		},
-		ObjectMeta: metav1.ObjectMeta{
-			Name:   networkaddonsnames.OPERATOR_CONFIG,
-			Labels: labels,
-		},
-		Spec: networkaddons.NetworkAddonsConfigSpec{
-			Multus:      &networkaddons.Multus{},
-			LinuxBridge: &networkaddons.LinuxBridge{},
-			KubeMacPool: &networkaddons.KubeMacPool{},
-		},
-	}
+func (r *ReconcileHyperConverged) ensureKubeVirtCommonTemplatebundle(instance *hcov1alpha1.HyperConverged, logger logr.Logger, request reconcile.Request) error {
+	kubevirtCommonTemplateBundle := newKubeVirtCommonTemplateBundleForCR(instance, request.Namespace)
+	return r.ensureResourceExists(instance, logger, request, kubevirtCommonTemplateBundle)
 }
 
-func newKubevirtCommonTemplateBundleForCR(cr *hcov1alpha1.HyperConverged) *sspv1.KubevirtCommonTemplatesBundle {
-	labels := map[string]string{
-		"app": cr.Name,
-	}
-	return &sspv1.KubevirtCommonTemplatesBundle{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      "common-templates-" + cr.Name,
-			Labels:    labels,
-			Namespace: "openshift",
-		},
-	}
+func (r *ReconcileHyperConverged) ensureKubeVirtNodeLabellerBundle(instance *hcov1alpha1.HyperConverged, logger logr.Logger, request reconcile.Request) error {
+	kubevirtNodeLabellerBundle := newKubeVirtNodeLabellerBundleForCR(instance, request.Namespace)
+	return r.ensureResourceExists(instance, logger, request, kubevirtNodeLabellerBundle)
 }
 
-func newKubevirtNodeLabellerBundleForCR(cr *hcov1alpha1.HyperConverged) *sspv1.KubevirtNodeLabellerBundle {
-	labels := map[string]string{
-		"app": cr.Name,
-	}
-	return &sspv1.KubevirtNodeLabellerBundle{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:   "node-labeller-" + cr.Name,
-			Labels: labels,
-		},
-	}
+func (r *ReconcileHyperConverged) ensureKubeVirtTemplateValidator(instance *hcov1alpha1.HyperConverged, logger logr.Logger, request reconcile.Request) error {
+	kubevirtTemplateValidator := newKubeVirtTemplateValidatorForCR(instance, request.Namespace)
+	return r.ensureResourceExists(instance, logger, request, kubevirtTemplateValidator)
 }
 
-func newKubevirtTemplateValidatorForCR(cr *hcov1alpha1.HyperConverged) *sspv1.KubevirtTemplateValidator {
-	labels := map[string]string{
-		"app": cr.Name,
-	}
-	return &sspv1.KubevirtTemplateValidator{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:   "template-validator-" + cr.Name,
-			Labels: labels,
-		},
-	}
+func (r *ReconcileHyperConverged) ensureKWebUI(instance *hcov1alpha1.HyperConverged, logger logr.Logger, request reconcile.Request) error {
+	kWebUI := newKWebUIForCR(instance, request.Namespace)
+	return r.ensureResourceExists(instance, logger, request, kWebUI)
 }
 
-func newKWebUIForCR(cr *hcov1alpha1.HyperConverged) *kwebuis.KWebUI {
-	labels := map[string]string{
-		"app": cr.Name,
+func (r *ReconcileHyperConverged) ensureResourceExists(instance *hcov1alpha1.HyperConverged, logger logr.Logger, request reconcile.Request, desiredRuntimeObj runtime.Object) error {
+	desiredMetaObj := desiredRuntimeObj.(metav1.Object)
```

Comment:
I'm not familiar with what you're doing in `desiredRuntimeObj.(metav1.Object)`.   Specially, how to the parenthesis around `metav1.Object` effect this call?

Reply:
The underlying object implements multiple interfaces. It comes in as runtime.Object. It is "cast" as metav1.Object here so that the Name and Namespace functions are accessible.